### PR TITLE
ARROW-6582: [R] Arrow to R fails with embedded nuls in strings

### PR DESCRIPTION
**Pull request by nealrichardson**

### CODE REVIEW

File: r/tests/testthat/test-Array.R [hunk 1]
```diff
@@ -616,6 +616,23 @@ test_that("Array$create() handles vector -> fixed size list arrays", {
...
+    as.raw(c(0x63, 0x61, 0x6d, 0x65, 0x72, 0x61)),
+    as.raw(c(0x74, 0x76))),
+    class = c("arrow_binary", "vctrs_vctr", "list"))
+  expect_error(rawToChar(raws[[3]]), "nul") # See?
+  array_with_nul <- Array$create(raws)$cast(utf8())
+  # In version 1.0.1, as.vector(array_with_nul) errors:
+  # Error in Array__as_vector(self) : embedded nul in string: 'ma\0n'
+  # On master (presumably this is a cpp11 thing) it does not error,
+  # but it terminates the string early:
+  # [1] "person" "woman"  "ma"     "camera" "tv"
```
Comment (dianaclarke) [inline r/tests/testthat/test-Array.R:633]: > # [1] "person" "woman"  "ma"     "camera" "tv"

😆

Comment (romainfrancois) [reply r/tests/testthat/test-Array.R:633]: 🙈

**Comment by romainfrancois** [comment]
based on @bkietz comment from https://github.com/apache/arrow/pull/8536 Converter_String::Ingest_some_nulls() could end with: 

```cpp
cpp11::unwind_protect([&] {
      if (array->null_count()) {
        // need to watch for nulls
        arrow::internal::BitmapReader null_reader(array->null_bitmap_data(),
                                                  array->offset(), n);
        for (int i = 0; i < n; i++, null_reader.Next()) {
          if (null_reader.IsSet()) {
            SET_STRING_ELT(data, start + i, unsafe_r_string(string_array->GetString(i)));
          } else {
            SET_STRING_ELT(data, start + i, NA_STRING);
          }
        }

      } else {
        for (int i = 0; i < n; i++) {
          SET_STRING_ELT(data, start + i, unsafe_r_string(string_array->GetString(i)));
        }
      }
    });
```

this builds on knowing that `GetString()` gives an utf8 string, and that we know its size, as opposed to what `cpp11::r_string(std::string)` does here: https://github.com/r-lib/cpp11/blob/master/inst/include/cpp11/r_string.hpp#L18 and uses unwind protection around the loop instead of the individual conversions. 

i.e. it assumes utf-8 but since it does not use the known size, it searches for `nul`: r_string(const std::string& data) : data_(safe[Rf_mkCharCE](data.c_str(), CE_UTF8)) {}

**Comment by romainfrancois** [comment]
@nealrichardson is the intent that we do get to the `embedded nul in string: 'ma\0n'` error ?

**Comment by jimhester** [comment]
Oh hmm, it is not on purpose, I think it was just copy pasted from the `const char*` implementation above. We should use `mkCharLenCE` for `std::string()` since we know the length.

That being said having an embedded `NULL` in any `CHARSXP` is not really going to work in R. Lots of places in the R codebase assumes strings are null terminated. I think the only way you can handle string data with embedded nulls in R is with raw vectors.

**Comment by nealrichardson** [comment]
Arguably failing is better than silently truncating, but that puts us back at the original user report. I see our options as:

1. Do nothing in the code; document a workflow of casting the `utf8` column to `binary`, then stripping out the `00`s yourself in R
2. In arrow_to_vector, if a nul is hit, downcast to binary and proceed with conversion (raise a warning instead of erroring)
3. Add a `skipNul`-like option (cf. https://stat.ethz.ch/R-manual/R-devel/library/base/html/readLines.html and others) and strip nulls if they're hit, raising a warning but keeping the result as a character vector.

**Comment by romainfrancois** [comment]
I think failing asap is better, either with the current code, or with an `unwind_protect()` outside the loop. 

```cpp
StringArrayType* string_array = static_cast<StringArrayType*>(array.get());
    auto unsafe_r_string = [](const std::string& s) {
      return Rf_mkCharCE(s.c_str(), CE_UTF8);
    };
    cpp11::unwind_protect([&] {
      if (array->null_count()) {
        // need to watch for nulls
        arrow::internal::BitmapReader null_reader(array->null_bitmap_data(),
                                                  array->offset(), n);
        for (int i = 0; i < n; i++, null_reader.Next()) {
          if (null_reader.IsSet()) {
            SET_STRING_ELT(data, start + i, unsafe_r_string(string_array->GetString(i)));
          } else {
            SET_STRING_ELT(data, start + i, NA_STRING);
          }
        }

      } else {
        for (int i = 0; i < n; i++) {
          SET_STRING_ELT(data, start + i, unsafe_r_string(string_array->GetString(i)));
        }
      }
    });

    return Status::OK();
```

**Comment by nealrichardson** [comment]
What you describe (including using GetView) is essentially what we now have on master: https://github.com/apache/arrow/blob/master/r/src/array_to_vector.cpp#L290-L321

The difference is that we moved back to `Rf_mkCharLenCE`, which is much faster because length is known and errors (correctly) when there are embedded nuls, instead of `Rf_mkCharCE`, which IIUC was the reason for the performance regression in #8356 and which was responsible for silently truncating at an embedded nul. 

If `Rf_mkCharLenCE` is what is raising the "embedded nul in string" error, then we have the option of catching that and falling back to a slower `skipNul` path to strip the nulls. If the error comes from somewhere else later, we may need to consider other options.

**Comment by romainfrancois** [comment]
It does look like `Rf_mkCharLenCE()` is generating the error: 

``` r
cpp11::cpp_eval('Rf_mkCharLenCE("camer\\0a", 6, CE_UTF8)')
#> Error in f(): embedded nul in string: 'camer\0'
```

<sup>Created on 2020-11-13 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by nealrichardson** [comment]
It would be good to get this resolved for 3.0. I pushed a naive fix: if `arrow.skip_nul = TRUE` (default FALSE, per base::readLines and base::scan), we go through a slow path and strip out nuls. A better solution (1) would check `arrow.skip_nul` outside of the loop (I could do this now but figure there's a smarter C/C++ way than I would have come up with); (2) would probably try the fast conversion path anyway and catch the exception and retry with the slow path, so even if arrow.skip_nul == true, we only do the slow path if there is a nul (I tried try/catch but didn't get it working right); and (3) it should raise a warning when it does strip a nul, as I believe readLines and scan do.

**Comment by bkietz** [comment]
@nealrichardson 1) I'll push an implementation of this 2) unfortunately, unwind_exceptions can't *really* be caught. They are used by cpp11 to get c++ stack unwinding correct but if one is currently in flight then the R runtime has already been informed that `stop()` was called. If the exception is caught and not rethrown, the R stack will continue to unwind (leaking c++ resources). 3) wilco

File: r/src/array_to_vector.cpp [hunk 2]
```diff
@@ -288,36 +290,104 @@ struct Converter_String : public Converter {
...
-            SET_STRING_ELT(data, start + i, r_string_from_view(string_array->GetView(i)));
-          } else {
-            SET_STRING_ELT(data, start + i, NA_STRING);
+        if (strip_out_nuls) {
+          for (int i = 0; i < n; i++) {
+            SET_STRING_ELT(data, start + i,
+                           r_string_from_view_strip_nul(string_array->GetView(i),
+                                                        &nul_was_stripped));
           }
+          return;
```
Comment (nealrichardson) [inline r/src/array_to_vector.cpp:308]: The main function return (L347) is `return Status::OK();` -- should these do the same? Or do I misunderstand what this return does?

File: r/src/array_to_vector.cpp [hunk 3]
```diff
@@ -288,36 +290,104 @@ struct Converter_String : public Converter {
...
+    }
+
     return Status::OK();
   }
 
   bool Parallel() const { return false; }
 
  private:
-  static SEXP r_string_from_view(const arrow::util::string_view& view) {
+  static SEXP r_string_from_view(arrow::util::string_view view) {
```
Comment (nealrichardson) [inline r/src/array_to_vector.cpp:353]: Why this change?

**Review by nealrichardson — COMMENTED** [review]
Thanks for doing this better :)

File: r/src/array_to_vector.cpp (hunk 2 above)
Comment (bkietz) [reply r/src/array_to_vector.cpp:308]: This `return` is breaking out of the lambda wrapped by `unwind_protect`; it doesn't make the parent scope return. If you prefer, we could use an `else` here

File: r/src/array_to_vector.cpp (hunk 3 above)
Comment (bkietz) [reply r/src/array_to_vector.cpp:353]: string_view is a trivially copyable structure so there's no need to protect it from copies with `const&`

File: r/src/array_to_vector.cpp (hunk 2 above)
Comment (nealrichardson) [reply r/src/array_to_vector.cpp:308]: no preference, just trying to understand

**Comment by johncassil** [comment]
@jimhester, @nealrichardson, @bkietz @dianaclarke @romainfrancois 

Just wanted to say thanks for working on this.  I reported it a long time ago and have just been periodically watching the developments slowly progress.   I'm excited to see that there will be a resolution!

Cheers!
